### PR TITLE
Feat: Add external models and loader to improve optimizer.

### PR DESCRIPTION
**Pull request by tobymao**
@z3z1ma

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -1129,6 +1137,30 @@ def create_python_model(
...
 
 
+def create_external_model(
+    name: str,
+    *,
+    dialect: t.Optional[str] = None,
+    path: Path = Path(),
+    **kwargs: t.Any,
+) -> Model:
+    """Creates a Seed model.
```
Comment (vchan) [inline sqlmesh/core/model/definition.py:1147]:
```suggestion
    """Creates an external model.
```

File: sqlmesh/core/schema_loader.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+from __future__ import annotations
+
+import typing as t
+from concurrent.futures import ThreadPoolExecutor
+from pathlib import Path
+
+from ruamel.yaml import YAML
+from sqlglot.dialects.dialect import DialectType as DialectType
```
Comment (vchan) [inline sqlmesh/core/schema_loader.py:8]:
```suggestion
from sqlglot.dialects.dialect import DialectType
```

File: sqlmesh/cli/main.py [hunk 3]
```diff
@@ -379,6 +379,14 @@ def rollback(obj: Context) -> None:
     obj.rollback()
 
 
+@cli.command("create_schema_file")
```
Comment (izeigerman) [inline sqlmesh/cli/main.py:382]: Can we somehow tie the command name back to the concept of external models? Like it's not at all clear what's a schema file and why should I care. Can we call it something like add_external_models? Or register_external_models?

File: sqlmesh/core/constants.py [hunk 4]
```diff
@@ -35,5 +35,6 @@
 HOOKS = "hooks"
 MACROS = "macros"
 MODELS = "models"
+SCHEMA = "schema.yaml"
```
Comment (izeigerman) [inline sqlmesh/core/constants.py:38]: [Nit] The constant name is misleading, can we call it `SCHEMA_YAML` instead?

File: sqlmesh/core/snapshot/definition.py [hunk 5]
```diff
@@ -252,8 +254,7 @@ class SnapshotTableInfo(PydanticModel, SnapshotInfoMixin, frozen=True):
     parents: t.Tuple[SnapshotId, ...]
     previous_versions: t.Tuple[SnapshotDataVersion, ...] = ()
     change_category: t.Optional[SnapshotChangeCategory]
-    is_materialized: bool
-    is_embedded_kind: bool
+    kind_name: ModelKindName
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/definition.py:257]: why can't we just call this `model_kind_name` here and eliminate a need for the property method in this class?

File: sqlmesh/core/snapshot/evaluator.py [hunk 6]
```diff
@@ -87,7 +87,8 @@ def evaluate(
                 tables / table clones should be used where applicable.
             kwargs: Additional kwargs to pass to the renderer.
         """
-        if snapshot.is_embedded_kind:
+        if snapshot.is_symbolic:
+            print(f"*** {snapshot.name} {snapshot.model_kind_name} {snapshot.is_symbolic}")
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/evaluator.py:91]: Shouldn't be here

File: sqlmesh/core/model/kind.py [hunk 7]
```diff
@@ -13,68 +13,85 @@
...
-    # TODO: Add support for snapshots
-    # SNAPSHOT = "SNAPSHOT"
-
-
-class ModelKind(PydanticModel):
-    name: ModelKindName
+class ModelKindMixin:
+    @property
+    def model_kind_name(self) -> ModelKindName:
+        raise NotImplementedError
```
Comment (izeigerman) [inline sqlmesh/core/model/kind.py:19]: `@abc.abstractmethod`?

File: tests/core/test_integration.py [hunk 8]
```diff
@@ -764,7 +764,7 @@ def validate_state_sync_environment(
 def validate_tables(snapshots: t.Iterable[Snapshot], context: Context) -> None:
     adapter = context.engine_adapter
     for snapshot in snapshots:
-        table_should_exist = not snapshot.is_embedded_kind
+        table_should_exist = not snapshot.is_embedded
```
Comment (izeigerman) [inline tests/core/test_integration.py:767]: Should this be `is_symbolic`?

File: sqlmesh/core/console.py [hunk 9]
```diff
@@ -470,12 +470,12 @@ def _snapshot_change_choices(
...
         if use_rich_formatting:
             indirect = f"[indirect]{indirect}[/indirect]"
-        if snapshot.is_view_kind:
+        if snapshot.is_view:
             choices = {
                 SnapshotChangeCategory.BREAKING: f"Update {direct} and backfill {indirect}",
                 SnapshotChangeCategory.NON_BREAKING: f"Update {direct} but don't backfill {indirect}",
             }
-        elif snapshot.is_embedded_kind:
+        elif snapshot.is_embedded:
```
Comment (izeigerman) [inline sqlmesh/core/console.py:478]: Should this be `is_symbolic`?

File: sqlmesh/core/snapshot/definition.py (hunk 5 above)
Comment (tobymao) [reply sqlmesh/core/snapshot/definition.py:257]: because then you can't override it, pydantic complains

File: sqlmesh/magics.py [hunk 10]
```diff
@@ -374,6 +374,12 @@ def migrate(self, line: str) -> None:
         self._context.migrate()
         self.display("Migration complete")
 
+    @magic_arguments()
+    @line_magic
+    def create_schema_file(self, line: str) -> None:
```
Comment (izeigerman) [inline sqlmesh/magics.py:379]: Ditto about the command name

File: tests/core/test_schema_loader.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+        SELECT * FROM sushi.raw_fruits
+    """,
+        )
+    )
+
+    sushi_context.upsert_model(model)
+
+    assert sushi_context.models["sushi.fruits"].columns_to_types == {
+        "id": exp.DataType.build("UNKNOWN"),
```
Comment (izeigerman) [inline tests/core/test_schema_loader.py:65]: Why types are UNKNOWN in this case?

File: sqlmesh/magics.py (hunk 10 above)
Comment (tobymao) [reply sqlmesh/magics.py]: create_external_models

File: tests/core/test_schema_loader.py (hunk 11 above)
Comment (tobymao) [reply tests/core/test_schema_loader.py:65]: thanks need to debug this

File: tests/core/test_schema_loader.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+        SELECT name FROM sushi.raw_fruits
+    """,
+        )
+    )
+
+    sushi_context.upsert_model(model)
+    sushi_context.create_schema_file()
+    assert sushi_context.models["sushi.fruits"].columns_to_types == {
+        "name": exp.DataType.build("UNKNOWN")
```
Comment (izeigerman) [inline tests/core/test_schema_loader.py:45]: Why `id` wasn't captured here? I can't find a logic which does the filtering based on columns used.

Comment (izeigerman) [reply tests/core/test_schema_loader.py:45]: Ahh, `sushi.fruits` is not external. Sorry, got confused.

File: tests/core/test_schema_loader.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        );
+
+        SELECT * FROM sushi.raw_fruits
+    """,
+        )
+    )
+
+    sushi_context.upsert_model(model)
+
+    assert sushi_context.models["sushi.fruits"].columns_to_types == {
```
Comment (izeigerman) [inline tests/core/test_schema_loader.py:64]: Can we also add a similar assert for the `sushi.raw_fruits` external model?

File: tests/core/test_schema_loader.py (hunk 12 above)
Comment (tobymao) [reply tests/core/test_schema_loader.py:45]: because sushi.fruits doesn't select id in the first version

File: setup.py [hunk 14]
```diff
@@ -44,7 +44,7 @@
         "requests",
         "rich",
         "ruamel.yaml",
-        "sqlglot~=13.0.1",
+        "sqlglot~=13.2.0",
```
Comment (GeorgeSittas) [inline setup.py:47]:
```suggestion
        "sqlglot~=13.2.1",
```

**Pull request merged**